Initial implementation of the QMP API code generator for #99

danderson opened the pull request:
It doesn't quite *do* anything yet, but it's already getting pretty big, so I figured I'd kick off some review.

Still missing: codegen for commands and events, and doing something useful with the docstrings.

danderson commented:
Oh good lord so many lint errors. Please consider those noted, working on fixing them.

danderson commented:
Fixed lint. I've kept my followup lint fixes as separate commits for ease of review, happy to squash them back into the parent commits if you want a cleaner history.

mdlayher commented:
This is amazing. I'll review this tomorrow!

mdlayher commented:
It appears you've committed a binary at `internal/qmp-gen/qmp-gen.test`.

mdlayher commented on internal/qmp-gen/parse_test.go:
```diff
@@ -0,0 +1,79 @@
+		in, out string
+	}{
+		{
+			// Basic translation
```
Add these comments as a field to the test table, `name string`.

mdlayher commented on internal/qmp-gen/parse_test.go:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	for i, test := range tests {
+		actual := strings.TrimSpace(string(pyToJSON([]byte(test.in))))
```
Once you've added a `name` field, use subtests!

```go
t.Run(tt.name, func(t *testing.T) {
    actual := strings.TrimSpace(string(pyToJSON([]byte(test.in))))
    // ...
})
```

Then you can also remove the `Case %d` from the log.

mdlayher commented on internal/qmp-gen/types.go:
```diff
@@ -0,0 +1,532 @@
+		fs := bytes.SplitN(part, []byte("\n{"), 2)
+		switch len(fs) {
+		default:
+			return nil, fmt.Errorf("Unexpected part of spec file %q: %s", path, string(part))
```
Error messages should begin with a lowercase letter.

danderson replied:
Done.

mdlayher commented on internal/qmp-gen/types.go:
```diff
@@ -0,0 +1,532 @@
+			}
+			ret[v.Name] = v
+		default:
+			return nil, fmt.Errorf("Unknown definition kind: %q", string(def.JSON))
```
Error messages should begin with a lowercase letter.

danderson replied:
Done.

mdlayher commented on internal/qmp-gen/types.go:
```diff
@@ -0,0 +1,532 @@
+	"udp":   true,
+	"uuid":  true,
+	"vm":    true,
+	"vnc":   true,
```
Add `acpi` to this list, and also alphabetize it for easy readability.

danderson replied:
Done.

benlemasurier replied:
`acpiost` as well

benlemasurier replied:
`fd`, `ftp`, `ftps`

benlemasurier replied:
`ppc`, `sparc`, `mips`, `vmdk`, `luks`

mdlayher commented:
Do we need `qmp/raw/spec.txt` committed when the spec can be fetched from a URL?

mdlayher commented on internal/qmp-gen/templates/main:
```diff
@@ -0,0 +1,24 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Let's slap a `// Generated using go:generate, do not edit` or similar comment in here in case anyone is browsing this file.

danderson replied:
Done.

benlemasurier commented on internal/qmp-gen/types.go:
```diff
@@ -0,0 +1,532 @@
+
+// getQAPI reads a QMP API spec file, from local disk or over HTTP(S).
+func getQAPI(path string) ([]byte, error) {
+	if path[0] == '/' {
```
`if path[0] == os.PathSeparator`

benlemasurier replied:
It may be a good idea to first `url.Parse()` here so we can handle relative filesystem paths.

danderson replied:
Done.

benlemasurier commented on internal/qmp-gen/main.go:
```diff
@@ -0,0 +1,77 @@
+	"io/ioutil"
+)
+
+const specURL = `https://raw.githubusercontent.com/danderson/qemu/master/qapi-schema.json`
```
any reason we shouldn't use `https://raw.githubusercontent.com/qemu/qemu/master/qapi-schema.json` instead?

danderson replied:
The official spec used to have inconsistencies that made it hard to parse. The patch I submitted upstream has been published (minus one regression which they're fixing), so now it's fine to use upstream again. Fixed.

danderson commented:
Removed the binary I mistakenly committed, and addressed your comments.

As for spec.txt - my reasoning is that when we hit bugs in the generation, it would be useful to have the corresponding spec "source" handy for comparison, rather than refer to the upstream spec which may have changed in the meantime. It also makes it easier to review future `go generate` updates, because you'll get the diff both to the source spec and the generated output side by side.

That said, I'm not married to the idea, happy to get rid of the spec.txt output if you think it's unnecessary.

mdlayher commented:
Regarding: https://github.com/digitalocean/go-qemu/pull/114#issuecomment-254292024.

I'm okay with keeping the spec in the repo.

danderson commented:
Comments addressed. Let me know if you'd like me to squash the commits, or if you prefer seeing the review history.

mdlayher commented:
I'm happy with this for sure. Let's keep the commit history because this change is so big. 

Has the latest generator been used to write out autogen.go? I'm mobile and can't tell.

LGTM.  Will let @benlemasurier do the final sign off.

danderson commented:
Yup, autogen.go is up to date with the generator code.

benlemasurier commented:
Excellent work!
